### PR TITLE
fix(windows): show bootstrap progress in setup

### PR DESCRIPTION
**Pull request by mcdurdin**
When Setup is starting, it can sometimes take a few seconds to show the main form because it is querying an online server and unzipping various files. This shows a small progress bar window so that the user doesn't wonder what is going on.

The progress bar is fairly basic: it has only 4 steps, and because all actions are happening on the main thread, does not animate well or smoothly. In my view, it is not worth redesigning into a multi-threaded process for the sake of this progress bar dialog.

Relates to #3689.

image

### CODE REVIEW

File: windows/src/desktop/setup/bootstrapmain.pas [hunk 1]
```diff
@@ -84,20 +83,45 @@ implementation
...
-procedure DoExtractOnly(FSilent: Boolean; const FExtractOnly_Path: string); forward;
-function CreateTempDir: string; forward;
-procedure RemoveTempDir(const path: string); forward;
-procedure ProcessCommandLine(var FPromptForReboot, FSilent, FForceOffline, FExtractOnly, FContinueSetup, FStartAfterInstall, FDisableUpgradeFrom6Or7Or8: Boolean; var FPackages, FExtractPath, FTier: string); forward;
-procedure SetExitVal(c: Integer); forward;
-function IsKeymanDesktop7Installed: string; forward;
-function IsKeymanDesktop8Installed: string; forward;
-function GetResourcesFromOnline(FSilent: Boolean; var FForceOffline: Boolean): Boolean; forward;
+type
+  TSetupBootstrap = class
```
Comment (mcdurdin) [inline windows/src/desktop/setup/bootstrapmain.pas:92]: In order to use the TfrmDownloadProgress dialog, we have to have a class to hold a callback function. This change just refactors all the above functions into the class with no functionality changes.

File: windows/src/desktop/setup/bootstrapmain.pas [hunk 2]
```diff
@@ -149,75 +176,25 @@ procedure Run;
...
-          // install, or point them to the old version of Keyman otherwise.
-          if CheckForOldVersionScenario then   // I4460
-            Exit;
-
-          TRunTools.CheckInstalledVersion(FInstallInfo.MsiInstallLocation);
-
-          // Looks for installation state for Keyman, and determines best packages for installation.
-          // If no .msi can be found, and Keyman is not installed, this will show/log an error and
-          // abort installation.
-          if not FInstallInfo.CheckMsiUpgradeScenarios then
```
Comment (mcdurdin) [inline windows/src/desktop/setup/bootstrapmain.pas:207]: This code is moved into `LocateResourcesCallback`.

**Comment by darcywong00** [comment]
In the screenshot, does clicking the "x" behave the same as "Cancel"?

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by mcdurdin** [comment]
> In the screenshot, does clicking the "x" behave the same as "Cancel"?

Yes

**Pull request merged**